src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs

```diff
@@ -33,21 +34,80 @@ namespace Microsoft.Identity.Client.Internal.Cache
     [DataContract]
     internal class RefreshTokenCacheItem : BaseTokenCacheItem
     {
+
         public RefreshTokenCacheItem()
         {
         }
 
-        public RefreshTokenCacheItem(string authority, string clientId, TokenResponse response) : base(authority, clientId, response)
+        public RefreshTokenCacheItem(string environment, string clientId, TokenResponse response) : base(clientId)
         {
             RefreshToken = response.RefreshToken;
+            Environment = environment;
+            PopulateIdentifiers(response);
+
+            User = User.Create(DisplayableId, Name, IdentityProvider,
+                GetUserIdentifier());
         }
 
+        [DataMember(Name = "environment")]
+        public string Environment { get; set; }
+
+        [DataMember(Name = "uid")]
+        public string Uid { get; set; }
+
+        [DataMember(Name = "utid")]
+        public string Utid { get; set; }
+
+        [DataMember(Name = "displayable_id")]
+        public string DisplayableId { get; internal set; }
+
+        [DataMember(Name = "name")]
+        public string Name { get; internal set; }
+
+        [DataMember(Name = "identity_provider")]
+        public string IdentityProvider { get; internal set; }
+
         [DataMember (Name = "refresh_token")]
         public string RefreshToken { get; set; }
 
-        public override TokenCacheKey GetTokenCacheKey()
+        public RefreshTokenCacheKey GetRefreshTokenItemKey()
+        {
+            return new RefreshTokenCacheKey(Environment, ClientId, GetUserIdentifier());
+        }
+
+        public void PopulateIdentifiers(TokenResponse response)
+        {
+            ClientInfo info = ClientInfo.Parse(response.ClientInfo);
+            IdToken idToken = IdToken.Parse(response.IdToken);
+            if (info != null)
+            {
+                Uid = info.UniqueIdentifier;
+                Utid = info.UniqueTenantIdentifier;
+            }
+            else
+            {
+                Uid = idToken.GetUniqueId();
+                Utid = idToken.TenantId;
+            }
+
+            DisplayableId = idToken.PreferredUsername;
+            Name = idToken.Name;
+            IdentityProvider = idToken.Issuer;
+        }
+
+        public sealed override string GetUserIdentifier()
+        {
+            return string.Format(CultureInfo.InvariantCulture, "{0}.{1}", MsalHelpers.EncodeToBase64Url(Uid),
+                MsalHelpers.EncodeToBase64Url(Utid));
+        }
+
+        // This method is called after the object 
+        // is completely deserialized.
+        [OnDeserialized]
+        void OnDeserialized(StreamingContext context)
         {
-            return new TokenCacheKey(null, null, ClientId, User.HomeObjectId);
+            User = User.Create(DisplayableId, Name, IdentityProvider,
```

Comment:
We can create User object from artifacts stored in the cache. Goal is to preserve things that are returned from the server. User is simply a representation in a particular fashion of the data stored.